src/index.ts

```diff
@@ -0,0 +1,66 @@
+import { createSocket, Socket } from 'dgram';
+import * as http from 'http';
+import { performance } from 'perf_hooks';
+
+type RequestOptions = http.RequestOptions;
+type RequestCallback = (response: http.IncomingMessage) => void;
+type RequestMethodArgs =
+  | [RequestOptions | string | URL, RequestCallback?]
+  | [string | URL, RequestOptions, RequestCallback?];
+
+
+type SendTelemetryFunction = (message: Buffer) => void
+
+interface AgentParams {
+  host?: string;
+  port?: number;
+  sendTelemetryFunction?: SendTelemetryFunction;
+}
+
+export class MetristAgent {
+  private _host: string;
+  private _port: number;
+  private _sendTelemetryFunction: SendTelemetryFunction;
+  private _socket: Socket;
+
+  constructor();
+  constructor(param: AgentParams);
+  constructor(param?: AgentParams) {
+    this._host = param?.host ?? process.env.METRIST_MONITORING_AGENT_HOST ?? '127.0.0.1';
+    this._port = param?.port ?? parseInt(process.env.METRIST_MONITORING_AGENT_PORT ?? '51712', 10);
+    this._sendTelemetryFunction = param?.sendTelemetryFunction ?? this.sendTelemetry;
+  }
+
+  connect() {
+    this._socket = createSocket('udp4');
+
+    const httpModule = require('http');
+    httpModule.request = this.patchHTTPMethod(httpModule, 'request');
+    httpModule.get = this.patchHTTPMethod(httpModule, 'get');
+
+    const httpsModule = require('https');
+    httpsModule.request = this.patchHTTPMethod(httpsModule, 'request');
+    httpsModule.get = this.patchHTTPMethod(httpsModule, 'get');
+  }
+
+  disconnect() {
+    this._socket.close();
+  }
+
+  private patchHTTPMethod(module: any, method: string) {
+    const httpMethod = module[method];
+    const agent = this;
+    return function(...args: RequestMethodArgs) {
+      const start = performance.now();
```

Comment:
This had a small bug since this was originally outside of the returned function, so every request was returning the time since the agent originally connected. Fixed, updated the test, and published the change to npm